File: src/memory/store.ts
import { createHash, randomUUID } from "node:crypto";
import { constants } from "node:fs";
import { link, lstat, mkdir, open, readdir, rename, rm } from "node:fs/promises";
import { basename, join, relative, resolve, sep } from "node:path";

const NOTES_RELATIVE = join(".agents", "notes");
const PENDING_RELATIVE = join(NOTES_RELATIVE, ".pending");
const CONSUMED_DIRECTORY = ".consumed";

export interface PendingNoteRecord {
  /** Path relative to cwd, using forward slashes. */
  path: string;
  content: string;
  /** SHA-256 of the UTF-8 file bytes. */
  hash: string;
}

export interface ConsolidatedNoteRecord {
  path: string;
  content: string;
  hash: string;
}

export interface ConsolidatedSaveReceipt {
  path: string;
  hash: string;
}

function errorCode(error: unknown): string | undefined {
  if (typeof error !== "object" || error === null || !("code" in error)) return undefined;
  const code = (error as { code?: unknown }).code;
  return typeof code === "string" ? code : undefined;
}

function digest(content: string): string {
  return createHash("sha256").update(content, "utf8").digest("hex");
}

function portableRelative(cwd: string, path: string): string {
  return relative(resolve(cwd), path).split(sep).join("/");
}

async function checkedDirectory(path: string, create: boolean): Promise<boolean> {
  try {
    const stat = await lstat(path);
    if (stat.isSymbolicLink()) throw new Error(`Refusing to traverse symlink: ${path}`);
    if (!stat.isDirectory()) throw new Error(`Managed memory path is not a directory: ${path}`);
    return true;
  } catch (error) {
    if (errorCode(error) !== "ENOENT") throw error;
    if (!create) return false;
    try {
      await mkdir(path, { mode: 0o700 });
    } catch (mkdirError) {
      if (errorCode(mkdirError) !== "EEXIST") throw mkdirError;
    }
    const stat = await lstat(path);
    if (stat.isSymbolicLink()) throw new Error(`Refusing to traverse symlink: ${path}`);
    if (!stat.isDirectory()) throw new Error(`Managed memory path is not a directory: ${path}`);
    return true;
  }
}

async function notesRoot(cwd: string, create: boolean): Promise<string | undefined> {
  const root = resolve(cwd);
  const agents = join(root, ".agents");
  if (!(await checkedDirectory(agents, create))) return undefined;
  const notes = join(agents, "notes");
  if (!(await checkedDirectory(notes, create))) return undefined;
  return notes;
}

async function ensureChildDirectories(root: string, segments: string[]): Promise<string> {
  let current = root;
  for (const segment of segments) {
    current = join(current, segment);
    await checkedDirectory(current, true);
  }
  return current;
}

function topicSegments(topic: string): string[] {
  if (topic === "" || topic === ".") return [];
  if (topic.includes("\\")) throw new Error("Topic must use forward-slash separators");
  const segments = topic.split("/");
  if (
    segments.some(
      (segment) => !segment || segment === "." || segment === ".." || segment.startsWith(".") || segment.includes("\0"),
    )
  ) {
    throw new Error(`Invalid memory topic: ${topic}`);
  }
  return segments;
}

function pendingPathFromRecord(cwd: string, recordPath: string): string {
  const normalized = recordPath.replaceAll("\\", "/");
  const expectedPrefix = ".agents/notes/.pending/";
  if (!normalized.startsWith(expectedPrefix)) throw new Error(`Invalid pending note path: ${recordPath}`);
  const name = normalized.slice(expectedPrefix.length);
  if (!name || name.includes("/") || basename(name) !== name || !name.endsWith(".md")) {
    throw new Error(`Invalid pending note path: ${recordPath}`);
  }
  return join(resolve(cwd), PENDING_RELATIVE, name);
}

async function readRegularFile(path: string): Promise<string> {
  const before = await lstat(path);
  if (before.isSymbolicLink()) throw new Error(`Refusing to read symlink: ${path}`);
  if (!before.isFile()) throw new Error(`Managed note is not a regular file: ${path}`);
  const handle = await open(path, constants.O_RDONLY | constants.O_NOFOLLOW);
  try {
    return await handle.readFile({ encoding: "utf8" });
  } finally {
    await handle.close();
  }
}

function consumptionReceipt(record: PendingNoteRecord): { name: string; payload: string } {
  // Include both the logical path and bytes: equal content at another filename is a distinct note.
  const name = `${digest(`${record.path}\0${record.content}`)}.json`;
  const payload = JSON.stringify({ path: record.path, hash: record.hash }) + "\n";
  return { name, payload };
}

async function hasConsumptionReceipt(consumed: string | undefined, record: PendingNoteRecord): Promise<boolean> {
  if (!consumed) return false;
  const receipt = consumptionReceipt(record);
  try {
    return (await readRegularFile(join(consumed, receipt.name))) === receipt.payload;
  } catch (error) {
    if (errorCode(error) === "ENOENT") return false;
    throw error;
  }
}

async function writeConsumptionReceipt(
  consumed: string,
  record: PendingNoteRecord,
  stillValid: () => boolean,
): Promise<boolean> {
  const receipt = consumptionReceipt(record);
  const destination = join(consumed, receipt.name);
  const temp = join(consumed, `.${receipt.name}.${randomUUID()}.tmp`);
  let handle: Awaited<ReturnType<typeof open>> | undefined;
  let tempCreated = false;
  try {
    handle = await open(temp, constants.O_WRONLY | constants.O_CREAT | constants.O_EXCL | constants.O_NOFOLLOW, 0o600);
    tempCreated = true;
    await handle.writeFile(receipt.payload, "utf8");
    await handle.sync();
    await handle.close();
    handle = undefined;

    // A hard link publishes the fully-written marker atomically and fails rather than
    // replacing any existing file (including a symlink).
    if (!stillValid()) return false;
    try {
      await link(temp, destination);
    } catch (error) {
      if (errorCode(error) !== "EEXIST" || !(await hasConsumptionReceipt(consumed, record))) throw error;
    }
    return true;
  } finally {
    await handle?.close().catch(() => undefined);
    if (tempCreated) await rm(temp, { force: true }).catch(() => undefined);
  }
}

/** Snapshot only direct, not-yet-consumed Markdown children of .agents/notes/.pending. */
export async function snapshotPendingNotes(cwd: string): Promise<PendingNoteRecord[]> {
  const notes = await notesRoot(cwd, false);
  if (!notes) return [];
  const pending = join(notes, ".pending");
  if (!(await checkedDirectory(pending, false))) return [];
  const consumedPath = join(notes, CONSUMED_DIRECTORY);
  const consumed = (await checkedDirectory(consumedPath, false)) ? consumedPath : undefined;

  const entries = await readdir(pending, { withFileTypes: true });
  const records: PendingNoteRecord[] = [];
  for (const entry of entries.sort((a, b) => a.name.localeCompare(b.name))) {
    if (!entry.name.endsWith(".md")) continue;
    const path = join(pending, entry.name);
    // Dirent information is not trusted: lstat/open enforce the no-symlink rule.
    const content = await readRegularFile(path);
    const record = { path: portableRelative(cwd, path), content, hash: digest(content) };
    if (!(await hasConsumptionReceipt(consumed, record))) records.push(record);
  }
  return records;
}

/**
 * Mark the supplied path/content snapshots as consumed without changing their mutable source files.
 * A later version at the same path has a different receipt identity and remains retryable.
 */
export async function consumePendingNotes(
  cwd: string,
  snapshot: readonly PendingNoteRecord[],
  stillValid: () => boolean = () => true,
): Promise<{ consumed: string[]; retained: string[] }> {
  if (!stillValid()) return { consumed: [], retained: snapshot.map((record) => record.path) };
  const notes = await notesRoot(cwd, true);
  if (!notes) return { consumed: [], retained: snapshot.map((record) => record.path) };
  const consumedDirectory = await ensureChildDirectories(notes, [CONSUMED_DIRECTORY]);

  const consumed: string[] = [];
  const retained: string[] = [];
  for (const record of snapshot) {
    // Refuse inconsistent caller-created records: snapshots produced here always satisfy this.
    if (!stillValid() || digest(record.content) !== record.hash) {
      retained.push(record.path);
      continue;
    }
    pendingPathFromRecord(cwd, record.path);
    if (await writeConsumptionReceipt(consumedDirectory, record, stillValid)) consumed.push(record.path);
    else retained.push(record.path);
  }
  return { consumed, retained };
}

/** Read a topic's index.md. The empty topic addresses .agents/notes/index.md. */
export async function readConsolidatedNote(cwd: string, topic = ""): Promise<ConsolidatedNoteRecord | undefined> {
  const notes = await notesRoot(cwd, false);
  if (!notes) return undefined;
  const segments = topicSegments(topic);
  let parent = notes;
  for (const segment of segments) {
    parent = join(parent, segment);
    if (!(await checkedDirectory(parent, false))) return undefined;
  }
  const path = join(parent, "index.md");
  try {
    const content = await readRegularFile(path);
    return { path: portableRelative(cwd, path), content, hash: digest(content) };
  } catch (error) {
    if (errorCode(error) === "ENOENT") return undefined;
    throw error;
  }
}

/**
 * Save one topic index while holding a cooperative per-topic lock.
 * Pass null to create a new index; pass the hash returned by readConsolidatedNote
 * when replacing one. The hash guard serializes callers of this API, but cannot
 * provide compare-and-swap guarantees against external writers that ignore the lock.
 */
export async function saveConsolidatedNote(
  cwd: string,
  topic: string,
  content: string,
  expectedHash: string | null,
): Promise<ConsolidatedSaveReceipt> {
  const segments = topicSegments(topic);
  const notes = await notesRoot(cwd, true);
  if (!notes) throw new Error("Unable to create managed notes directory");
  const parent = await ensureChildDirectories(notes, segments);
  const path = join(parent, "index.md");
  const lockPath = join(parent, ".index.md.lock");
  let lockHandle: Awaited<ReturnType<typeof open>>;
  try {
    lockHandle = await open(
      lockPath,
      constants.O_WRONLY | constants.O_CREAT | constants.O_EXCL | constants.O_NOFOLLOW,
      0o600,
    );
  } catch (error) {
    if (errorCode(error) === "EEXIST") {
      throw new Error(`Consolidated note is locked by another cooperative writer: ${portableRelative(cwd, path)}`);
    }
    throw error;
  }

  const temp = join(parent, `.index.md.${randomUUID()}.tmp`);
  let tempCreated = false;
  try {
    let existing: string | undefined;
    try {
      existing = await readRegularFile(path);
    } catch (error) {
      if (errorCode(error) !== "ENOENT") throw error;
    }
    if (existing === undefined ? expectedHash !== null : expectedHash === null || digest(existing) !== expectedHash) {
      throw new Error(`Consolidated note changed or already exists: ${portableRelative(cwd, path)}`);
    }

    const handle = await open(
      temp,
      constants.O_WRONLY | constants.O_CREAT | constants.O_EXCL | constants.O_NOFOLLOW,
      0o600,
    );
    tempCreated = true;
    try {
      await handle.writeFile(content, "utf8");
      await handle.sync();
    } finally {
      await handle.close();
    }
    await rename(temp, path);
    tempCreated = false;
    return { path: portableRelative(cwd, path), hash: digest(content) };
  } finally {
    if (tempCreated) await rm(temp, { force: true }).catch(() => undefined);
    await lockHandle.close().catch(() => undefined);
    await rm(lockPath, { force: true }).catch(() => undefined);
  }
}

/** Verify that a receipt still describes the bytes saved at its managed topic index. */
export async function verifyConsolidatedSaveReceipt(cwd: string, receipt: ConsolidatedSaveReceipt): Promise<boolean> {
  const prefix = ".agents/notes/";
  const normalized = receipt.path.replaceAll("\\", "/");
  if (
    !normalized.startsWith(prefix) ||
    normalized.includes("/.pending/") ||
    (!normalized.endsWith("/index.md") && normalized !== ".agents/notes/index.md")
  ) {
    return false;
  }
  const topicPath = normalized === ".agents/notes/index.md" ? "" : normalized.slice(prefix.length, -"/index.md".length);
  try {
    const record = await readConsolidatedNote(cwd, topicPath);
    return record?.path === normalized && record.hash === receipt.hash;
  } catch {
    return false;
  }
}

File: expt-go/internal/memory/store.go
// Package memory stores project-local agent memory as ordinary Markdown files.
//
// The package does not schedule agents, summarize content, or take a project lock.
// A coordinator explicitly prepares a consolidation, calls a provider child, and
// commits the result after the child has terminated successfully.
package memory

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

const (
	MaxPendingFiles      = 256
	MaxPendingFileBytes  = 1 << 20
	MaxPendingTotalBytes = 8 << 20

	maxMarkerBytes = 1024
)

var (
	ErrChanged = errors.New("memory content changed")
)

type PendingNote struct {
	Path    string // slash-separated, relative to the project root
	Content string
	SHA256  string
}

type Note struct {
	Path, Content, SHA256 string
}

type SaveReceipt struct {
	Path   string `json:"path"`
	SHA256 string `json:"sha256"`
}

type ConsumeResult struct {
	Consumed []string
	Retained []string
}

type Event struct {
	Operation string
	RunID     string
	Message   string
	Count     int
}

type Logger func(Event)

type Store struct {
	root string
	log  Logger
}

func New(root string, logger Logger) (*Store, error) {
	if root == "" {
		return nil, errors.New("memory: project root is required")
	}
	absolute, err := filepath.Abs(root)
	if err != nil {
		return nil, fmt.Errorf("memory: resolve project root: %w", err)
	}
	return &Store{root: absolute, log: logger}, nil
}

func (s *Store) Root() string { return s.root }
func (s *Store) emit(e Event) {
	if s.log != nil {
		s.log(e)
	}
}

func digest(b []byte) string { sum := sha256.Sum256(b); return hex.EncodeToString(sum[:]) }
func slashRelative(root, path string) string {
	r, _ := filepath.Rel(root, path)
	return filepath.ToSlash(r)
}

func checkedDir(path string, create bool) (bool, error) {
	info, err := os.Lstat(path)
	if err == nil {
		if info.Mode()&os.ModeSymlink != 0 {
			return false, fmt.Errorf("memory: refusing to traverse symlink: %s", path)
		}
		if !info.IsDir() {
			return false, fmt.Errorf("memory: managed path is not a directory: %s", path)
		}
		return true, nil
	}
	if !errors.Is(err, os.ErrNotExist) {
		return false, err
	}
	if !create {
		return false, nil
	}
	if err := os.Mkdir(path, 0700); err != nil && !errors.Is(err, os.ErrExist) {
		return false, err
	}
	info, err = os.Lstat(path)
	if err != nil {
		return false, err
	}
	if info.Mode()&os.ModeSymlink != 0 {
		return false, fmt.Errorf("memory: refusing to traverse symlink: %s", path)
	}
	if !info.IsDir() {
		return false, fmt.Errorf("memory: managed path is not a directory: %s", path)
	}
	return true, nil
}

func (s *Store) notesRoot(create bool) (string, bool, error) {
	agents := filepath.Join(s.root, ".agents")
	ok, err := checkedDir(agents, create)
	if err != nil || !ok {
		return "", false, err
	}
	notes := filepath.Join(agents, "notes")
	ok, err = checkedDir(notes, create)
	if err != nil || !ok {
		return "", false, err
	}
	return notes, true, nil
}

func ensureDirs(root string, segments []string) (string, error) {
	current := root
	for _, segment := range segments {
		current = filepath.Join(current, segment)
		if _, err := checkedDir(current, true); err != nil {
			return "", err
		}
	}
	return current, nil
}

func topicSegments(topic string) ([]string, error) {
	if topic == "" || topic == "." {
		return nil, nil
	}
	if strings.Contains(topic, "\\") {
		return nil, fmt.Errorf("memory: topic must use forward-slash separators")
	}
	parts := strings.Split(topic, "/")
	for _, p := range parts {
		if p == "" || p == "." || p == ".." || strings.HasPrefix(p, ".") || strings.ContainsRune(p, 0) {
			return nil, fmt.Errorf("memory: invalid topic %q", topic)
		}
	}
	return parts, nil
}

func readRegular(path string, maximum int64) ([]byte, error) {
	before, err := os.Lstat(path)
	if err != nil {
		return nil, err
	}
	if before.Mode()&os.ModeSymlink != 0 {
		return nil, fmt.Errorf("memory: refusing to read symlink: %s", path)
	}
	if !before.Mode().IsRegular() {
		return nil, fmt.Errorf("memory: managed note is not a regular file: %s", path)
	}
	if maximum >= 0 && before.Size() > maximum {
		return nil, fmt.Errorf("memory: managed note exceeds the %d-byte read limit: %s", maximum, path)
	}
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()
	var reader io.Reader = f
	if maximum >= 0 {
		reader = io.LimitReader(f, maximum+1)
	}
	b, err := io.ReadAll(reader)
	if err != nil {
		return nil, err
	}
	if maximum >= 0 && int64(len(b)) > maximum {
		return nil, fmt.Errorf("memory: managed note exceeds the %d-byte read limit: %s", maximum, path)
	}
	return b, nil
}

func (s *Store) SnapshotPending() ([]PendingNote, error) {
	notes, ok, err := s.notesRoot(false)
	if err != nil || !ok {
		return nil, err
	}
	pending := filepath.Join(notes, ".pending")
	ok, err = checkedDir(pending, false)
	if err != nil || !ok {
		return nil, err
	}
	entries, err := os.ReadDir(pending)
	if err != nil {
		return nil, err
	}
	names := make([]string, 0)
	for _, entry := range entries {
		if !strings.HasSuffix(entry.Name(), ".md") {
			continue
		}
		names = append(names, entry.Name())
		if len(names) > MaxPendingFiles {
			return nil, fmt.Errorf("memory: pending memory exceeds the %d-file limit", MaxPendingFiles)
		}
	}
	sort.Strings(names)
	consumedPath := filepath.Join(notes, ".consumed")
	consumedOK, err := checkedDir(consumedPath, false)
	if err != nil {
		return nil, err
	}
	var total int64
	out := make([]PendingNote, 0, len(names))
	for _, name := range names {
		path := filepath.Join(pending, name)
		b, err := readRegular(path, MaxPendingFileBytes)
		if err != nil {
			return nil, err
		}
		total += int64(len(b))
		if total > MaxPendingTotalBytes {
			return nil, fmt.Errorf("memory: pending memory exceeds the %d-byte aggregate limit", MaxPendingTotalBytes)
		}
		record := PendingNote{Path: slashRelative(s.root, path), Content: string(b), SHA256: digest(b)}
		used, err := s.hasMarker(consumedPath, consumedOK, record)
		if err != nil {
			return nil, err
		}
		if !used {
			out = append(out, record)
		}
	}
	s.emit(Event{Operation: "snapshot", Message: "pending snapshot created", Count: len(out)})
	return out, nil
}

func marker(record PendingNote) (string, []byte) {
	identity := digest([]byte(record.Path+"\x00"+record.Content)) + ".json"
	payload, _ := json.Marshal(struct {
		Path   string `json:"path"`
		SHA256 string `json:"hash"`
	}{record.Path, record.SHA256})
	return identity, append(payload, '\n')
}

func (s *Store) hasMarker(dir string, exists bool, record PendingNote) (bool, error) {
	if !exists {
		return false, nil
	}
	name, payload := marker(record)
	got, err := readRegular(filepath.Join(dir, name), maxMarkerBytes)
	if errors.Is(err, os.ErrNotExist) {
		return false, nil
	}
	if err != nil {
		return false, err
	}
	return string(got) == string(payload), nil
}

func validPendingPath(path string) bool {
	p := filepath.ToSlash(path)
	const prefix = ".agents/notes/.pending/"
	if !strings.HasPrefix(p, prefix) {
		return false
	}
	name := strings.TrimPrefix(p, prefix)
	return name != "" && !strings.Contains(name, "/") && strings.HasSuffix(name, ".md")
}

func (s *Store) Consume(snapshot []PendingNote, stillValid func() bool) (ConsumeResult, error) {
	result := ConsumeResult{}
	valid := func() bool { return stillValid == nil || stillValid() }
	if !valid() {
		for _, r := range snapshot {
			result.Retained = append(result.Retained, r.Path)
		}
		return result, nil
	}
	notes, _, err := s.notesRoot(true)
	if err != nil {
		return result, err
	}
	dir, err := ensureDirs(notes, []string{".consumed"})
	if err != nil {
		return result, err
	}
	for _, record := range snapshot {
		if !valid() || digest([]byte(record.Content)) != record.SHA256 {
			result.Retained = append(result.Retained, record.Path)
			continue
		}
		if !validPendingPath(record.Path) {
			return result, fmt.Errorf("memory: invalid pending note path: %s", record.Path)
		}
		name, payload := marker(record)
		tmp, err := os.CreateTemp(dir, "."+name+".*.tmp")
		if err != nil {
			return result, err
		}
		tmpName := tmp.Name()
		cleanup := func() { tmp.Close(); os.Remove(tmpName) }
		if err = tmp.Chmod(0600); err == nil {
			_, err = tmp.Write(payload)
		}
		if err == nil {
			err = tmp.Sync()
		}
		if closeErr := tmp.Close(); err == nil {
			err = closeErr
		}
		if err != nil {
			cleanup()
			return result, err
		}
		if !valid() {
			cleanup()
			result.Retained = append(result.Retained, record.Path)
			continue
		}
		destination := filepath.Join(dir, name)
		err = os.Link(tmpName, destination)
		if errors.Is(err, os.ErrExist) {
			var same bool
			same, err = s.hasMarker(dir, true, record)
			if err == nil && !same {
				err = errors.New("memory: conflicting consumption marker")
			}
		}
		cleanup()
		if err != nil {
			return result, err
		}
		result.Consumed = append(result.Consumed, record.Path)
	}
	s.emit(Event{Operation: "consume", Message: "pending snapshots marked consumed", Count: len(result.Consumed)})
	return result, nil
}

func (s *Store) Read(topic string) (*Note, error) {
	segments, err := topicSegments(topic)
	if err != nil {
		return nil, err
	}
	notes, ok, err := s.notesRoot(false)
	if err != nil || !ok {
		return nil, err
	}
	parent := notes
	for _, segment := range segments {
		parent = filepath.Join(parent, segment)
		ok, err = checkedDir(parent, false)
		if err != nil || !ok {
			return nil, err
		}
	}
	path := filepath.Join(parent, "index.md")
	b, err := readRegular(path, -1)
	if errors.Is(err, os.ErrNotExist) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return &Note{Path: slashRelative(s.root, path), Content: string(b), SHA256: digest(b)}, nil
}

func (s *Store) Save(topic, content string, expectedSHA256 *string) (SaveReceipt, error) {
	segments, err := topicSegments(topic)
	if err != nil {
		return SaveReceipt{}, err
	}
	notes, _, err := s.notesRoot(true)
	if err != nil {
		return SaveReceipt{}, err
	}
	parent, err := ensureDirs(notes, segments)
	if err != nil {
		return SaveReceipt{}, err
	}
	path := filepath.Join(parent, "index.md")
	existing, readErr := readRegular(path, -1)
	if readErr != nil && !errors.Is(readErr, os.ErrNotExist) {
		return SaveReceipt{}, readErr
	}
	if errors.Is(readErr, os.ErrNotExist) {
		if expectedSHA256 != nil {
			return SaveReceipt{}, fmt.Errorf("%w: %s does not exist", ErrChanged, slashRelative(s.root, path))
		}
	} else if expectedSHA256 == nil || digest(existing) != *expectedSHA256 {
		return SaveReceipt{}, fmt.Errorf("%w: %s", ErrChanged, slashRelative(s.root, path))
	}
	tmp, err := os.CreateTemp(parent, ".index.md.*.tmp")
	if err != nil {
		return SaveReceipt{}, err
	}
	tmpName := tmp.Name()
	defer os.Remove(tmpName)
	if err = tmp.Chmod(0600); err == nil {
		_, err = tmp.WriteString(content)
	}
	if err == nil {
		err = tmp.Sync()
	}
	if closeErr := tmp.Close(); err == nil {
		err = closeErr
	}
	if err != nil {
		return SaveReceipt{}, err
	}
	if err = os.Rename(tmpName, path); err != nil {
		return SaveReceipt{}, err
	}
	receipt := SaveReceipt{Path: slashRelative(s.root, path), SHA256: digest([]byte(content))}
	s.emit(Event{Operation: "save", Message: "consolidated index saved", Count: 1})
	return receipt, nil
}

func (s *Store) VerifySave(receipt SaveReceipt) bool {
	p := filepath.ToSlash(receipt.Path)
	const prefix = ".agents/notes/"
	if !strings.HasPrefix(p, prefix) || strings.Contains(p, "/.pending/") || !(p == prefix+"index.md" || strings.HasSuffix(p, "/index.md")) {
		return false
	}
	topic := ""
	if p != prefix+"index.md" {
		topic = strings.TrimSuffix(strings.TrimPrefix(p, prefix), "/index.md")
	}
	note, err := s.Read(topic)
	return err == nil && note != nil && note.Path == p && note.SHA256 == receipt.SHA256
}

func randomID() (string, error) {
	var b [16]byte
	if _, err := rand.Read(b[:]); err != nil {
		return "", err
	}
	return hex.EncodeToString(b[:]), nil
}
